src/osd/OSD.cc

```diff
@@ -7242,11 +7240,11 @@ void OSD::dispatch_context_transaction(PG::RecoveryCtx &ctx, PG *pg,
                                        ThreadPool::TPHandle *handle)
 {
   if (!ctx.transaction->empty()) {
-    ctx.on_applied->add(new ObjectStore::C_DeleteTransaction(ctx.transaction));
     int tr = store->queue_transaction(
       pg->osr.get(),
-      ctx.transaction, ctx.on_applied, ctx.on_safe, NULL,
+      std::move(*ctx.transaction), ctx.on_applied, ctx.on_safe, NULL,
       TrackedOpRef(), handle);
+    delete (ctx.transaction);
```

Comment:
This needs some changes and in fact we can do this kind of optimization lot of other places specially those are in io/recovery path..
IMO, we should stage it as already the delta of changes for this is quite big, what do you say ?
Also, we need some soak time to see it is working without any problem in al the cases..